Firestore/core/src/model/values.cc

```diff
@@ -0,0 +1,477 @@
+                    ServerTimestamps::GetLocalWriteTime(right));
+    case kTypeOrderString:
+      return nanopb::MakeStringView(left.string_value) ==
+             nanopb::MakeStringView(right.string_value);
+    case kTypeOrderBlob:
+      return CompareBlobs(left, right) == ComparisonResult::Same;
+    case kTypeOrderReference:
+      return nanopb::MakeStringView(left.reference_value) ==
+             nanopb::MakeStringView(right.reference_value);
+    case kTypeOrderGeoPoint:
+      return left.geo_point_value.latitude == right.geo_point_value.latitude &&
+             left.geo_point_value.longitude == right.geo_point_value.longitude;
+    case kTypeOrderArray:
+      return ArrayEquals(left, right);
+    case kTypeOrderMap:
+      return ObjectEquals(left, right);
+    default:
+      HARD_FAIL("Invalid type value: %s", left_type);
+  }
+}
+
+bool Values::NumberEquals(
+    const firebase::firestore::google_firestore_v1_Value& left,
+    const firebase::firestore::google_firestore_v1_Value& right) {
+  if (left.which_value_type == google_firestore_v1_Value_integer_value_tag &&
+      right.which_value_type == google_firestore_v1_Value_integer_value_tag) {
+    return left.integer_value == right.integer_value;
+  } else if (left.which_value_type ==
+                 google_firestore_v1_Value_double_value_tag &&
+             right.which_value_type ==
+                 google_firestore_v1_Value_double_value_tag) {
+    return util::DoubleBitwiseEquals(left.double_value, right.double_value);
+  }
+  return false;
+}
+
+bool Values::ArrayEquals(
+    const firebase::firestore::google_firestore_v1_Value& left,
+    const firebase::firestore::google_firestore_v1_Value& right) {
+  const google_firestore_v1_ArrayValue& left_array = left.array_value;
+  const google_firestore_v1_ArrayValue& right_array = right.array_value;
+
+  if (left_array.values_count != right_array.values_count) {
+    return false;
+  }
+
+  for (size_t i = 0; i < left_array.values_count; ++i) {
+    if (!Equals(left_array.values[i], right_array.values[i])) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+bool Values::ObjectEquals(
+    const firebase::firestore::google_firestore_v1_Value& left,
+    const firebase::firestore::google_firestore_v1_Value& right) {
+  google_firestore_v1_MapValue left_map = left.map_value;
+  google_firestore_v1_MapValue right_map = right.map_value;
+
+  if (left_map.fields_count != right_map.fields_count) {
+    return false;
+  }
+
+  // Create a map of field names to index for one of the maps. This is then used
+  // look up the corresponding value for the other map's fields.
+  std::unordered_map<std::string, size_t> key_to_value_index;
+  for (size_t i = 0; i < left_map.fields_count; ++i) {
+    key_to_value_index.emplace(nanopb::MakeString(left_map.fields[i].key), i);
+  }
+
+  for (size_t i = 0; i < right_map.fields_count; ++i) {
+    const std::string& key = nanopb::MakeString(right_map.fields[i].key);
```

Comment:
Should also be StringView because it is only used to look up.

Reply:
This doesn't seem to compile since I am looking up a std::string in the map. I tried changing the map to absl::string_view, but it did what I expected it to do and failed to compile since there is no hash function. I also think that using a string_view as a map key might be a bit of an antipattern but you or Costa can probably explain better.

```
std::unordered_map<absl::string_view, size_t, absl::Hash<absl::string_view>> key_to_value_index;


Undefined symbols for architecture x86_64:
  "absl::lts_2020_02_25::hash_internal::CityHash64(char const*, unsigned long)", referenced from:
      absl::lts_2020_02_25::hash_internal::CityHashState::CombineContiguousImpl(unsigned long long, unsigned char const*, unsigned long, std::__1::integral_constant<int, 8>) in libfirestore_core.a(values.cc.o)
  "absl::lts_2020_02_25::hash_internal::CityHashState::CombineLargeContiguousImpl64(unsigned long long, unsigned char const*, unsigned long)", referenced from:
      absl::lts_2020_02_25::hash_internal::CityHashState::CombineContiguousImpl(unsigned long long, unsigned char const*, unsigned long, std::__1::integral_constant<int, 8>) in libfirestore_core.a(values.cc.o)
  "absl::lts_2020_02_25::hash_internal::CityHashState::kSeed", referenced from:
      absl::lts_2020_02_25::hash_internal::CityHashState::CityHashState() in libfirestore_core.a(values.cc.o)
```

Reply:
Ah, my bad, I think you are right. We definitely need someone to review my review :-)

Reply:
Are you trying to use `flat_hash_map` (judging by `container_internal::FlatHashMapPolicy`)? Can you try adding `absl::flat_hash_map` to the dependencies?

Reply:
I added both absl_hash and absl_container to the dependencies (which should pull in the hash map), but it still does not seem to be working: https://gist.github.com/schmidt-sebastian/6145cab8862e4a72ce7e949fa4b7ef92